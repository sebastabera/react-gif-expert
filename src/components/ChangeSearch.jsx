import { useState } from "react";
import PropTypes from 'prop-types';

export const ChangeSearch = ({onNewSearch}) => {
  
    const [inputValue, setinputValue] = useState('')

    const onInputChange = ({target}) => {
        setinputValue(target.value);
    };

    const onSubmit = (event) => {
        event.preventDefault();
        if(inputValue.trim().length <= 1) return;

        onNewSearch(inputValue.trim());
        setinputValue('');
    };

    return (
        <form onSubmit={onSubmit} aria-label="form" >
            <input 
                type="text" 
                placeholder="User ID" 
                value={inputValue}
                onChange={onInputChange} 
                required />
        </form>
       
    )
}

ChangeSearch.propTypes = {
    onNewSearch: PropTypes.func.isRequired
}