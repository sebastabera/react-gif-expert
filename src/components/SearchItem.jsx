import PropTypes from 'prop-types';

export const SearchItem = ({courseName, url}) => {
  
  return (
    <div className='card'>
        <img src={url} alt={courseName} width='400'/>
        <p>{courseName}</p>
    </div>
  )
}

SearchItem.propTypes = {
  courseName: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
