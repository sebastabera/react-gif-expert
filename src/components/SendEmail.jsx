import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';

import { useState } from "react";
import PropTypes from 'prop-types';
import { sendEmail } from "../helpers/sendEmail";

export const SendEmail = ({user, products}) => {

    const [emailValue, setEmailValue] = useState('');

    const onInputChange = ({target}) => {
        setEmailValue(target.value);
    };

    const notifySuccess = () => toast.success(`ðŸ“¨ Email sent with recommendation for the user ${user}`, {
            position: "top-right",
            autoClose: 5000,
            hideProgressBar: false,
            closeOnClick: true,
            pauseOnHover: true,
            draggable: true,
            progress: undefined,
            theme: "light",
        });
    const notifyError = (error) => toast.error(`${error}`);

    const onSubmit = async (event) => {
        event.preventDefault();
        if(emailValue.trim().length <= 1) return;
        try {
            const {statusCode, body } = await sendEmail(emailValue, products);
            if(statusCode >= 300) {
                const {error} = JSON.parse(body);
                notifyError(error);
                return;
            }
            notifySuccess();
            setEmailValue('');  
        } catch (error) {
            notifyError(error);
        }
    };

    return (
        <>
            <h3>Sending email with recomendations</h3>
            <form onSubmit={onSubmit} aria-label="form" >
                <input 
                    type="email" 
                    placeholder="Email" 
                    value={emailValue}
                    onChange={onInputChange} 
                    required />
            </form>
            <ToastContainer />
        </>
       
    )
}

SendEmail.propTypes = {
    user: PropTypes.string.isRequired, 
    products: PropTypes.array.isRequired
}