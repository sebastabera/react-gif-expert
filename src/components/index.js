export * from './ChangeSearch';
export * from './SearchResultGrid';
export * from './SearchItem';