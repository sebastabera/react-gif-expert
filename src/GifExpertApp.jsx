import {useState} from 'react';
import { AddCategory, GifGrid } from './components';

export const GifExpertApp = () => {

    const [categories, setCategories] = useState(['One Punch', 'Valorant']);

    const onAddCategory = (newCategory) => {
        if(categories.includes(newCategory)) return;
        setCategories([newCategory, ...categories]);
        //setCategories(cat => [...cat, 'Dragon Ball']);
    };

  return (
    <>

        <h1>GifExpertApp</h1>

        <AddCategory 
            onNewCategory={value => onAddCategory(value)}/>

        {
            categories.map( category => (
                <GifGrid key={category} category={category}/>
            )
        )}

    </>
  )
}
