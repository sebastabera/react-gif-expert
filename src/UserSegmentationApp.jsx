import {useState} from 'react';
import { ChangeSearch, SearchResultGrid } from './components';

export const UserSegmentationApp = () => {

    const [search, setSearch] = useState('');
    const [recomendations, setRecomendations] = useState([]);

    const onChangeSearch = (newSearch) => {
        if(search === newSearch) return;
        setSearch(newSearch);
        //setCategories(cat => [...cat, 'Dragon Ball']);
    };
    console.log(search);

  return (
    <>

        <h1>User Segmentation App</h1>

        <ChangeSearch 
            onNewSearch={value => onChangeSearch(value)}/>

        {search !== '' && 
            <SearchResultGrid key={search} search={search}/>
        }

    </>
  )
}
