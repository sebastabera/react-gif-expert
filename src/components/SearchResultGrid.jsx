import PropTypes from 'prop-types';
import { useFetchProducts } from "../hooks/useFetchProducts";
import { SearchItem } from "./SearchItem";

export const SearchResultGrid = ({search}) => {

    const {products, isLoading} = useFetchProducts(search);

    return (
        <>
            
            <h3>Recomendaciones para: {search}</h3>
            {
                isLoading && ( <h2>Cargando</h2> )
            }
            <div className="card-grid">
                {products.map((product) => (
                    <SearchItem key={product.id} {...product} />
                ))}
            </div>
            
        </>
    )
}

SearchResultGrid.propTypes = {
    search: PropTypes.string.isRequired
}
