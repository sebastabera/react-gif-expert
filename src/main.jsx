import React from 'react'
import ReactDOM from 'react-dom/client'
import { UserSegmentationApp } from './UserSegmentationApp'
import './styles.css';

ReactDOM.createRoot(document.getElementById('root')).render(
  <React.StrictMode>
    <UserSegmentationApp />
  </React.StrictMode>,
);
