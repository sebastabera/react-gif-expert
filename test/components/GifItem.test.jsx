import { render, screen } from "@testing-library/react"
import { GifItem } from "../../src/components"

describe('Pruebas en <GifItem />', () => { 

    const title = 'Saitama';
    const url = 'https://one-punch.com/saitama'
    
    test('Debe hacer match con el snapshop', () => { 
        
        const {container} = render(<GifItem title={title} url={url} />);
        expect (container).toMatchSnapshot();
     })

    test('Debe de mostrar la imagen con el URL y el ALT indicado', () => { 
        
        render(<GifItem title={title} url={url} />);
        //screen.debug();

        const {alt, src} = screen.getByRole('img');

        expect(src).toBe(url);
        expect(alt).toBe(title);
     });

     test('Debe de mostrar el titulo en el componente', () => { 
        
        render(<GifItem title={title} url={url} />);
        expect(screen.getByText(title)).toBeTruthy();

      })

 })