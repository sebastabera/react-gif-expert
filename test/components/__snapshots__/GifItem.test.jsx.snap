// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifItem /> Debe hacer match con el snapshop 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Saitama"
      src="https://one-punch.com/saitama"
    />
    <p>
      Saitama
    </p>
  </div>
</div>
`;
